### PR TITLE
Support soxr custom recipes.

### PR DESCRIPTION
**Pull request by bitkeeper**
MPD uses soxr with prefined resample recipes. Soxr also support defining a recipe your self.
This commit will support a custom recipe by changing the existing quality setting to "custom".

The same structs as the predefined recipes uses can now be set by hand.

This will make the following settings available:

- precision 16|20|24|28|32 bits, example "28"
- phase_response - 0-50, example "45"
- passband_end - used bandwidth of source 80-99.7%, example "99.7.0"
- stopband_begin - anti aliasing 100.0+%, example "100".
- attenuation - signal reduciton in dB's, 0-30. example "3.0".
- flags "0" - additional bitmask with extra settings

The data is set in the structs soxr_quality_spec and soxr_io_spec (found in soxr.h).

### CODE REVIEW

File: src/pcm/SoxrResampler.cxx [hunk 1]
```diff
@@ -29,7 +29,7 @@
 #include <cassert>
 
 #include <string.h>
-
+#include <math.h>
```
Comment (MaxKellermann) [inline src/pcm/SoxrResampler.cxx:32]: Use C++ headers for new code, i.e. `<cmath>`

File: src/pcm/SoxrResampler.cxx [hunk 2]
```diff
@@ -80,20 +88,110 @@ soxr_parse_quality(const char *quality) noexcept
 	return SOXR_INVALID_RECIPE;
 }
 
+static unsigned 
+soxr_parse_precision(unsigned value) {
+	switch(value) {
+		case 16:
```
Comment (MaxKellermann) [inline src/pcm/SoxrResampler.cxx:94]: Bad coding style, see https://www.musicpd.org/doc/html/developer.html#code-style
This problem affects most of your code!

File: src/pcm/SoxrResampler.cxx [hunk 3]
```diff
@@ -80,20 +88,110 @@ soxr_parse_quality(const char *quality) noexcept
...
+		default:
+			throw FormatInvalidArgument("soxr converter invalid precision : %d [16|20|24|28|32]",
+					    value);
+			break;
+	}
+	return value;
+}
+
+static double
+soxr_parse_phase_respons(signed value) {
```
Comment (MaxKellermann) [inline src/pcm/SoxrResampler.cxx:109]: Why make this `signed` if negative values are not allowed?
And why is this integer if it implicitly casts to floating point?

File: src/pcm/SoxrResampler.cxx [hunk 4]
```diff
@@ -80,20 +88,110 @@ soxr_parse_quality(const char *quality) noexcept
...
+	}
+
+	return value;
+}
+
+static double
+soxr_parse_passband_end(const char* svalue) {
+	char* endptr;
+	double value=strtod (svalue, &endptr);
+	if(*endptr) {
```
Comment (MaxKellermann) [inline src/pcm/SoxrResampler.cxx:122]: This allows parsing an empty string.

File: src/pcm/SoxrResampler.cxx [hunk 5]
```diff
@@ -39,9 +39,16 @@ static constexpr unsigned long SOXR_DEFAULT_RECIPE = SOXR_HQ;
  */
 static constexpr unsigned long SOXR_INVALID_RECIPE = -1;
 
+/**
+ * Special value for custom recipe.
+ */
+static constexpr unsigned long SOXR_CUSTOM_RECIPE = -2;
+
+static soxr_io_spec_t soxr_io;
```
Comment (MaxKellermann) [inline src/pcm/SoxrResampler.cxx:47]: I don't like the variable name; this is for storing the custom settings and is unused otherwise, but the name doesn't tell that.

File: src/pcm/SoxrResampler.cxx [hunk 6]
```diff
@@ -109,14 +207,19 @@ SoxrPcmResampler::Open(AudioFormat &af, unsigned new_sample_rate)
 	assert(audio_valid_sample_rate(new_sample_rate));
 
 	soxr_error_t e;
+	soxr_io_spec_t* p_soxr_io = nullptr;
+	if(soxr_io.scale != 0) {
```
Comment (MaxKellermann) [inline src/pcm/SoxrResampler.cxx:211]: This is a bad check for "using custom settings". This happens to work in practice, but depends on implicit zero initialization of a floating point in a global variable - which is obscure. Better be explicit here about checking `SOXR_CUSTOM_RECIPE`.

File: src/pcm/SoxrResampler.cxx [hunk 7]
```diff
@@ -109,14 +207,19 @@ SoxrPcmResampler::Open(AudioFormat &af, unsigned new_sample_rate)
...
 	soxr = soxr_create(af.sample_rate, new_sample_rate,
 			   af.channels, &e,
-			   nullptr, &soxr_quality, &soxr_runtime);
+			   p_soxr_io, &soxr_quality, &soxr_runtime);
 	if (soxr == nullptr)
 		throw FormatRuntimeError("soxr initialization has failed: %s",
 					 e);
 
 	FormatDebug(soxr_domain, "soxr engine '%s'", soxr_engine(soxr));
+	FormatDebug(soxr_domain, "soxr precision=%0.2f, phase_response=%0.2f, passband_end=%0.2f, stopband_begin=%0.2f scale=%0.2f", soxr_quality.precision, soxr_quality.phase_response, soxr_quality.passband_end, soxr_quality.stopband_begin, soxr_io.scale  );
```
Comment (MaxKellermann) [inline src/pcm/SoxrResampler.cxx:222]: This is logged even if the global variable is never initialized.

File: src/pcm/SoxrResampler.cxx (hunk 2 above)
Comment (bitkeeper) [reply src/pcm/SoxrResampler.cxx:94]: Was not sure to if I should follow the code-style or the style of the already existing code surrounding the code change. Ok clear always use the code-style.

File: src/pcm/SoxrResampler.cxx (hunk 5 above)
Comment (bitkeeper) [reply src/pcm/SoxrResampler.cxx:47]: Is soxr_io_custom_recipe better ?

File: src/pcm/SoxrResampler.cxx (hunk 6 above)
Comment (bitkeeper) [reply src/pcm/SoxrResampler.cxx:211]: I will make sure that the  struct is initialized, else an additional variable would be required.

File: src/pcm/SoxrResampler.cxx (hunk 7 above)
Comment (bitkeeper) [reply src/pcm/SoxrResampler.cxx:222]: What do you prefer leaving the scale out or an if with two different  FormatDebug statements ?

**Comment by neheb** [comment]
I suggest running through git clang-format

**Comment by bitkeeper** [comment]
@neheb Thanks for the tip.

**Comment by neheb** [comment]
You have 3 separate commits. Need to run it through all of them.

Actually...you should probably squash all of the commits.

File: src/pcm/SoxrResampler.cxx [hunk 8]
```diff
@@ -28,20 +28,35 @@
...
 
 static constexpr unsigned long SOXR_DEFAULT_RECIPE = SOXR_HQ;
 
+/**
+ * The SOXR_RESET_ON_CLEAR flag makes it possible to jump in the position off the current 
+ * track. Applied when using a custom recipe.
+ */
+static constexpr unsigned long SOXR_RESET_ON_CLEAR =
+	2147483648; // 2^31 defined in internal.h of soxr as RESET_ON_CLEAR
+
```
Comment (MaxKellermann) [inline src/pcm/SoxrResampler.cxx:44]: This wasn't in your first version of the PR - and I don't like using undocumented features with internal flags. This can break any time.

File: src/pcm/SoxrResampler.cxx [hunk 9]
```diff
@@ -51,8 +66,8 @@ static constexpr struct {
 	{ SOXR_MQ, "medium" },
 	{ SOXR_LQ, "low" },
 	{ SOXR_QQ, "quick" },
-	{ SOXR_INVALID_RECIPE, nullptr }
-};
+	{ SOXR_CUSTOM_RECIPE, "custom"},
+	{ SOXR_INVALID_RECIPE, nullptr}};
```
Comment (MaxKellermann) [inline src/pcm/SoxrResampler.cxx:70]: Don't reformat existing code in an unrelated commit

File: src/pcm/SoxrResampler.cxx [hunk 10]
```diff
@@ -80,20 +95,114 @@ soxr_parse_quality(const char *quality) noexcept
...
+	}
+
+	return double(value);
+}
+
+static double
+SoxrParsePassbandEnd(const char *svalue) {
+	char *endptr;
+	double value = strtod(svalue, &endptr);
+	if (strlen(svalue) == 0 || *endptr) {
```
Comment (MaxKellermann) [inline src/pcm/SoxrResampler.cxx:129]: This is a clumsy check - three things are wrong here:
- it use `strlen()` to check if a string is empty, but you could better use `*svalue == 0`
- if you check `svalue`, you better do it before calling `strtod()`, because this doesn't depend on `strtod()`
- ... but even that would be complicated; the canonical way to rule out a parsing error here is checking `svalue == endptr` because that will tell you whether anything has been parsed at all

File: src/pcm/SoxrResampler.cxx [hunk 11]
```diff
@@ -80,20 +95,114 @@ soxr_parse_quality(const char *quality) noexcept
...
+			throw FormatInvalidArgument(
+				"soxr converter passband_end value not a number: %s", svalue);
+	}
+
+	if (value < 1 || value > 100) {
+			throw FormatInvalidArgument(
+				"soxr converter invalid passband_end : %d (1-100%)", value);
```
Comment (MaxKellermann) [inline src/pcm/SoxrResampler.cxx:136]: Two bugs in this format string:
- using `%d` but passing a `double`
- unescaped `%`

File: src/pcm/SoxrResampler.cxx [hunk 12]
```diff
@@ -80,20 +95,114 @@ soxr_parse_quality(const char *quality) noexcept
...
+	}
+
+	return value / 100.0;
+}
+
+static double
+SoxrParseStopbandBegin(const char *svalue) {
+	char *endptr;
+	double value = strtod(svalue, &endptr);
+	if (strlen(svalue) == 0 || *endptr) {
```
Comment (MaxKellermann) [inline src/pcm/SoxrResampler.cxx:146]: See above

File: src/pcm/SoxrResampler.cxx [hunk 13]
```diff
@@ -80,20 +95,114 @@ soxr_parse_quality(const char *quality) noexcept
...
+			throw FormatInvalidArgument(
+				"soxr converter stopband_begin value not a number: %s", svalue);
+	}
+
+	if (value < 100 || value > 199) {
+			throw FormatInvalidArgument(
+				"soxr converter invalid stopband_begin : %d (100-150%)", value);
```
Comment (MaxKellermann) [inline src/pcm/SoxrResampler.cxx:153]: See above

File: src/pcm/SoxrResampler.cxx [hunk 14]
```diff
@@ -80,20 +95,114 @@ soxr_parse_quality(const char *quality) noexcept
...
+	}
+
+	return value / 100.0;
+}
+
+static double
+SoxrParseAttenuation(const char *svalue) {
+	char *endptr;
+	double value = strtod(svalue, &endptr);
+	if (*endptr) {
```
Comment (MaxKellermann) [inline src/pcm/SoxrResampler.cxx:163]: `endptr==svalue` missing

File: src/pcm/SoxrResampler.cxx [hunk 15]
```diff
@@ -80,20 +95,114 @@ soxr_parse_quality(const char *quality) noexcept
...
+			throw FormatInvalidArgument(
+				"soxr converter attenuation value not a number: %s", svalue);
+	}
+
+	if (value < 0 || value > 30) {
+			throw FormatInvalidArgument(
+				"soxr converter invalid attenuation : %d (0-30)%)", value);
```
Comment (MaxKellermann) [inline src/pcm/SoxrResampler.cxx:170]: See above

File: src/pcm/SoxrResampler.cxx [hunk 16]
```diff
@@ -80,20 +95,114 @@ soxr_parse_quality(const char *quality) noexcept
...
+SoxrParseAttenuation(const char *svalue) {
+	char *endptr;
+	double value = strtod(svalue, &endptr);
+	if (*endptr) {
+			throw FormatInvalidArgument(
+				"soxr converter attenuation value not a number: %s", svalue);
+	}
+
+	if (value < 0 || value > 30) {
+			throw FormatInvalidArgument(
```
Comment (MaxKellermann) [inline src/pcm/SoxrResampler.cxx:169]: Bad indent

File: src/pcm/SoxrResampler.cxx [hunk 17]
```diff
@@ -109,14 +218,32 @@ SoxrPcmResampler::Open(AudioFormat &af, unsigned new_sample_rate)
 	assert(audio_valid_sample_rate(new_sample_rate));
 
 	soxr_error_t e;
+	soxr_io_spec_t* p_soxr_io = nullptr;
+	if(soxr_io_custom_recipe.scale != 0) {
```
Comment (MaxKellermann) [inline src/pcm/SoxrResampler.cxx:222]: This is still not an explicit check for "custom recipe".
You initialized the global variable, but it is not guaranteed per soxr API documentation. Looking and the soxr source code, this looks completely wrong, because `soxr_io_spec()` initializes `scale=1`. Did you ever test this?
And: comparing floating-point values with `==` is conceptually broken. You should avoid this.

File: src/pcm/SoxrResampler.cxx (hunk 8 above)
Comment (bitkeeper) [reply src/pcm/SoxrResampler.cxx:44]: Ok then I hide it by call `soxr_quality_spec` to init `soxr_quality`.

File: src/pcm/SoxrResampler.cxx (hunk 17 above)
Comment (bitkeeper) [reply src/pcm/SoxrResampler.cxx:222]: Yes you are right about the scale=1 vs 0 mixing up linear vs db.
Will add static variable soxr_use_custom_recipe for explit use.

**Comment by bitkeeper** [comment]
Will squash the commits when no more changes are required.

File: src/pcm/SoxrResampler.cxx [hunk 18]
```diff
@@ -51,7 +60,8 @@ static constexpr struct {
 	{ SOXR_MQ, "medium" },
 	{ SOXR_LQ, "low" },
 	{ SOXR_QQ, "quick" },
-	{ SOXR_INVALID_RECIPE, nullptr }
+	{ SOXR_CUSTOM_RECIPE, "custom"},
+	{ SOXR_INVALID_RECIPE, nullptr}
```
Comment (MaxKellermann) [inline src/pcm/SoxrResampler.cxx:64]: Whitespace change

File: src/pcm/SoxrResampler.cxx [hunk 19]
```diff
@@ -80,20 +90,118 @@ soxr_parse_quality(const char *quality) noexcept
...
 }
 
+static unsigned
+SoxrParsePrecision(unsigned value) {
+	switch (value) {
+		case 16:
+		case 20:
+		case 24:
+		case 28:
+		case 32:
```
Comment (MaxKellermann) [inline src/pcm/SoxrResampler.cxx:100]: Bad indent

File: src/pcm/SoxrResampler.cxx [hunk 20]
```diff
@@ -80,20 +90,118 @@ soxr_parse_quality(const char *quality) noexcept
...
+			break;
+		default:
+			throw FormatInvalidArgument(
+				"soxr converter invalid precision : %d [16|20|24|28|32]", value);
+			break;
```
Comment (MaxKellermann) [inline src/pcm/SoxrResampler.cxx:105]: Unreachable `break`

File: src/pcm/SoxrResampler.cxx [hunk 21]
```diff
@@ -80,20 +90,118 @@ soxr_parse_quality(const char *quality) noexcept
...
+	}
+	return value;
+}
+
+static double
+SoxrParsePhaseRespons(unsigned value) {
```
Comment (MaxKellermann) [inline src/pcm/SoxrResampler.cxx:111]: Typo in function name?

File: src/pcm/SoxrResampler.cxx [hunk 22]
```diff
@@ -80,20 +90,118 @@ soxr_parse_quality(const char *quality) noexcept
...
+	}
+
+	return double(value);
+}
+
+static double
+SoxrParsePassbandEnd(const char *svalue) {
+	char *endptr;
+	double value = strtod(svalue, &endptr);
+	if (svalue == endptr) {
```
Comment (MaxKellermann) [inline src/pcm/SoxrResampler.cxx:124]: This doesn't check for garbage after the parsed number

File: src/pcm/SoxrResampler.cxx [hunk 23]
```diff
@@ -80,20 +90,118 @@ soxr_parse_quality(const char *quality) noexcept
...
 	if (recipe == SOXR_INVALID_RECIPE) {
 		assert(quality_string != nullptr);
-
 		throw FormatRuntimeError("unknown quality setting '%s' in line %d",
-					 quality_string, block.line);
-	}
-
-	soxr_quality = soxr_quality_spec(recipe, 0);
-
+					quality_string, block.line);
```
Comment (MaxKellermann) [inline src/pcm/SoxrResampler.cxx:182]: Whitespace change

File: src/pcm/SoxrResampler.cxx [hunk 24]
```diff
@@ -80,20 +90,118 @@ soxr_parse_quality(const char *quality) noexcept
...
+	return 1 / pow(10, value / 10.0);
+}
+
 void
 pcm_resample_soxr_global_init(const ConfigBlock &block)
 {
 	const char *quality_string = block.GetBlockValue("quality");
 	unsigned long recipe = soxr_parse_quality(quality_string);
+	soxr_use_custom_recipe = recipe == SOXR_CUSTOM_RECIPE;
+	soxr_io_custom_recipe = soxr_io_spec(SOXR_FLOAT32_I, SOXR_FLOAT32_I);
```
Comment (MaxKellermann) [inline src/pcm/SoxrResampler.cxx:177]: Initialize this variable only if this is really a custom recipe - move the initialization to the `if (recipe == SOXR_CUSTOM_RECIPE) {` block.

**Review by MaxKellermann — CHANGES_REQUESTED** [review]

File: src/pcm/SoxrResampler.cxx (hunk 22 above)
Comment (bitkeeper) [reply src/pcm/SoxrResampler.cxx:124]: When no number could be parsed the a expection is rased and if the number isn't within the expected range also.
Can you clarify ? You mean garbage as like a string with "1.23 ABC" where only the first part is parsed ?

File: src/pcm/SoxrResampler.cxx (hunk 19 above)
Comment (bitkeeper) [reply src/pcm/SoxrResampler.cxx:100]: The preview in github shows more indent then it really is in the file itself. Please look at the raw file.
The `case` has two times the indent as for `switch`.

File: src/pcm/SoxrResampler.cxx (hunk 22 above)
Comment (MaxKellermann) [reply src/pcm/SoxrResampler.cxx:124]: > You mean garbage as like a string with "1.23 ABC" where only the first part is parsed ?

Yes. Such a string should raise an error, because MPD should not silently ignore such garbage.

Usually, after `strto*()`, you check for errors using `endptr == s || *endptr != 0`. This checks whether anything was parsed at all, and whether everything was parsed.

File: src/pcm/SoxrResampler.cxx (hunk 19 above)
Comment (MaxKellermann) [reply src/pcm/SoxrResampler.cxx:100]: >  The `case` has two times the indent as for `switch`.

That's the problem. It should have the same indent.
Coding style is a matter of taste, of course - and your indentation isn't absolutely wrong, just doesn't match the coding style used by the rest of MPD.

File: src/pcm/SoxrResampler.cxx [hunk 25]
```diff
@@ -80,20 +90,117 @@ soxr_parse_quality(const char *quality) noexcept
...
+		throw FormatInvalidArgument(
+			"soxr converter attenuation value not a number: %s", svalue);
+	}
+
+	if (value < 0 || value > 30) {
+		throw FormatInvalidArgument(
+			"soxr converter invalid attenuation : %s (0-30dB))", svalue);
+	}
+
+	return 1 / pow(10, value / 10.0);
```
Comment (MaxKellermann) [inline src/pcm/SoxrResampler.cxx:167]: You included `<cmath>`, and this header doesn't provide `pow()`, but only `std::pow()`.

**Comment by MaxKellermann** [comment]
I think this looks good now. Please rebase & fold.

**Comment by MaxKellermann** [comment]
Oh and please fix all those trailing whitespace. When I look at your code in my editor (and not GitHub), my screen turns red due to code style warnings...

File: src/pcm/SoxrResampler.cxx (hunk 25 above)
Comment (bitkeeper) [reply src/pcm/SoxrResampler.cxx]: hmm strange that it did compile ... something else is pulling it in,  but I have changed it and added the namespace.

Comment (MaxKellermann) [reply src/pcm/SoxrResampler.cxx]: Usually, C++ library implementations implement `<cmath>` by including the good old C header `<math.h>` and then copy symbols over to `namespace std`, so this *usually* works. But strictly, it is a mistake to assume that the symbol without namespace qualification is available. This may break any time, and I want MPD to be as correct as possible.

**Comment by bitkeeper** [comment]
> Oh and please fix all those trailing whitespace. When I look at your code in my editor (and not GitHub), my screen turns red due to code style warnings...

Hmm I don't see any trailing whitespace in my editors ?

**Comment by bitkeeper** [comment]
> I think this looks good now. Please rebase & fold.

done

**Comment by MaxKellermann** [comment]
9aa432c07838256536c39b00d14c04462862b4fc
I fixed the whitespace errors for you.
I did not fix the author name. It's not some obscure GitHub auto-generated address.
